pkg/operator/ceph/controller/version.go

```diff
@@ -47,3 +58,79 @@ func GetImageVersion(cephCluster cephv1.CephCluster) (*cephver.CephVersion, erro
 
 	return nil, errors.New("attempt to determine ceph version for the current cluster image timed out")
 }
+
+// DetectCephVersion loads the ceph version from the image and checks that it meets the version requirements to
+// run in the cluster
+func DetectCephVersion(rookImage, namespace, jobName string, ownerInfo *k8sutil.OwnerInfo, clientset kubernetes.Interface, cephClusterSpec *cephv1.ClusterSpec) (*version.CephVersion, error) {
+	cephImage := cephClusterSpec.CephVersion.Image
+	logger.Infof("detecting the ceph image version for image %s...", cephImage)
+	versionReporter, err := cmdreporter.New(
+		clientset,
+		ownerInfo,
+		jobName,
+		jobName,
+		namespace,
+		[]string{"ceph"},
+		[]string{"--version"},
+		rookImage,
+		cephImage,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set up ceph version job")
+	}
+
+	job := versionReporter.Job()
+	job.Spec.Template.Spec.ServiceAccountName = "rook-ceph-cmd-reporter"
+
+	// Apply the same placement for the ceph version detection as the mon daemons except for PodAntiAffinity
+	cephv1.GetMonPlacement(cephClusterSpec.Placement).ApplyToPodSpec(&job.Spec.Template.Spec)
+	job.Spec.Template.Spec.Affinity.PodAntiAffinity = nil
+
+	stdout, stderr, retcode, err := versionReporter.Run(detectCephVersionTimeout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to complete ceph version job")
+	}
+	if retcode != 0 {
+		return nil, errors.Errorf(`ceph version job returned failure with retcode %d. `+
+			`stdout: %s. `+
+			`stderr: %s`,
+			retcode,
+			stdout,
+			stderr,
+		)
+	}
+
+	version, err := cephver.ExtractCephVersion(stdout)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to extract ceph version")
+	}
+	logger.Infof("detected ceph image version: %q", version)
+	return version, nil
+}
+
+func CurrentAndDesiredCephVersion(rookImage, namespace, jobName string, ownerInfo *k8sutil.OwnerInfo, context *clusterd.Context, cephClusterSpec *cephv1.ClusterSpec, clusterInfo *cephclient.ClusterInfo) (*version.CephVersion, *version.CephVersion, error) {
```

Comment:
I thought they were going to share version info somehow to avoid running the job so many times, but maybe I'm just missing it. When testing on this PR, it is concerning to see jobs run for each separate controller to detect the same thing. It's a heavyweight action.

Reply:
I agree it's heavyweight but all controllers are independent and they don't really share any sort of state despite the clusters context. I didn't have time to think about a design. Maybe this is something we can consider as a follow-up and discuss with @BlaineEXE?

Reply:
As discussed, I'm really just suggesting an optimization. We can follow up with a separate PR, but certainly low priority.

Reply:
Definitely can be in a different PR. 

I think all of the CRs other than `CephObjectStore` implicitly or explicitly require the Ceph cluster to be created before they can be created. Similarly, I think it's okay to wait to update CRDs until after the Ceph cluster reconcile loop has detected the updated Ceph version.

One thing I wonder if we could do is have two different contexts, one for the CephCluster reconciler and the other for all other reconcilers. If the CephCluster reconciler is canceled, the other can be canceled also. But if the CephCluster reconciler determines that there is a Ceph update needed, could it cancel the other context after it updates its status with the latest Ceph version info?

Another strategy we could take that would surely work would be to keep a ConfigMap that has the CephCluster reconciler's most recently detected Ceph version in it. All other reconcilers could watch for changes to that configmap and run a new reconcile if they see changes.

Or we could watch for changes to the CephCluster as we do today and run a reconcile if the status.cephVersion is updated.

Reply:
> Definitely can be in a different PR.
> 
> I think all of the CRs other than `CephObjectStore` implicitly or explicitly require the Ceph cluster to be created before they can be created. Similarly, I think it's okay to wait to update CRDs until after the Ceph cluster reconcile loop has detected the updated Ceph version.
> 
> One thing I wonder if we could do is have two different contexts, one for the CephCluster reconciler and the other for all other reconcilers. If the CephCluster reconciler is canceled, the other can be canceled also. But if the CephCluster reconciler determines that there is a Ceph update needed, could it cancel the other context after it updates its status with the latest Ceph version info?

We cannot have different contexts. The manager accepts a single context and passes it down to all the controllers.
So today if the main context is canceled then all the managers are.
Your proposal is not achievable since each controller runs independently.

> 
> Another strategy we could take that would surely work would be to keep a ConfigMap that has the CephCluster reconciler's most recently detected Ceph version in it. All other reconcilers could watch for changes to that configmap and run a new reconcile if they see changes.

This only works for updates, if the operator is restarted, all controllers will start and all of them will attempt to update at the same time. We don't want that, that's why we always compare the running version and the desired version from the image.

> 
> Or we could watch for changes to the CephCluster and run a reconcile if the status.cephVersion is updated.

Same problem as above. Additionally, it takes some amount of time to update the cephcluster, at least the monitors. So checking the CR status is not reliable. All the time the child will think nothing is being upgraded since the status hasn't been updated yet.

I'm still convinced that the cmdreporter could appear expensive but is definitely the most reliable way.